dpo/aya_dpo_gen.py

```diff
@@ -0,0 +1,222 @@
+@step(inputs=["targets", "generation"], outputs=["generations"])
+def CombineAyaAndModelResponse(
+    inputs: StepInput,
+) -> StepOutput:
+    """A step to combine the Aya and model responses and add the response sources."""
+    for input in inputs:
+        input["generations"] = [input["targets"], input["generation"]]
+        input["response_source"] = ["aya", MODEL_ID]
+    yield inputs
+
+
+def update_argilla_dataset_with_metadata(
+    dataset_name: str,
+    workspace: str,
+    hub_dataset: Dataset,
+    metadata_keys: List[str] = None,
+):
+    """Update an Argilla dataset with metadata from the dataset."""
+    # by default, we add the predicted generation language and response source
+    if metadata_keys is None:
+        metadata_keys = ["predicted_generation_language", "response_source"]
+    dataset = rg.FeedbackDataset.from_argilla(
+        dataset_name,
+        workspace=workspace,
+    )
+
+    metadata_values = [hub_dataset[key] for key in metadata_keys]
+
+    if any(len(values) != len(dataset.records) for values in metadata_values):
+        raise ValueError(
+            f"Number of metadata values does not match the number of records ({len(dataset.records)})"
+        )
+
+    modified_records = []
+    for record, *metadata in zip(dataset.records, *metadata_values):
+        for key, value in zip(metadata_keys, metadata):
+            record.metadata[key] = value
+        modified_records.append(record)
+    dataset.update_records(modified_records)
+
+
+with Pipeline(name="generate-dpo-responses") as pipeline:
+    # Load the dataset from the Hugging Face Hub
+    load_hub_dataset = LoadHubDataset(
+        name="load_dataset",
+        output_mappings={"inputs": "instruction"},
+    )
+    # Generate responses using the model
+    text_generation = TextGeneration(
+        name="text_generation",
+        llm=InferenceEndpointsLLM(
+            base_url=INFERENCE_ENDPOINTS_URL,
+            tokenizer_id=MODEL_ID,
+            model_display_name=MODEL_ID,
+            api_key=HUGGINGFACE_TOKEN,
+        ),
+        input_batch_size=INPUT_BATCH_SIZE,
+        output_mappings={"model_name": "generation_model"},
+        num_generations=2,
+    )
+    load_hub_dataset.connect(text_generation)
+    language_prediction = LanguagePredict(name="language_prediction")
+    text_generation.connect(language_prediction)
+    combine_columns = CombineAyaAndModelResponse(
+        name="combine_columns",
+    )
+    language_prediction.connect(combine_columns)
+    to_argilla = PreferenceToArgilla(
+        name="to_argilla",
+        dataset_name=ARGILLA_DATASET_NAME,
+        api_url=ARGILLA_SPACE_URL,
+        api_key=ARGILLA_API_KEY,
+        dataset_workspace=ARGILLA_WORKSPACE_NAME,
+        num_generations=2,
+        input_mappings={
+            "instruction": "instruction",
+            "generations": "generations",
+        },
+    )
+    combine_columns.connect(to_argilla)
```

Comment:
Also I believe we can rewrite this to add a custom `Step` that creates a `TermsMetadataProperty` instead for the languages, so that those can be easily filtered as per the screen recording below!

https://github.com/huggingface/data-is-better-together/assets/36760800/6cb7908a-3c36-4fb4-831d-56fd26fdc9ef

Reply:
That would be great! Sometimes the language switch is correctly following the prompt but it would be useful to be able to dig into those examples quickly to check that.